yew-function-components-macro/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+use proc_macro2::{Span, TokenStream};
+use quote::{quote, quote_spanned, ToTokens};
+use syn::parse::{Parse, ParseStream};
+use syn::spanned::Spanned;
+use syn::{
+    parse_macro_input, Attribute, Block, FnArg, Ident, Item, ItemFn, ReturnType, Type, Visibility,
+};
+
+struct FunctionComponent {
+    block: Box<Block>,
+    props_type: Box<Type>,
+    arg: FnArg,
+    vis: Visibility,
+    attrs: Vec<Attribute>,
+    name: Ident,
+    return_type: Box<Type>,
+}
+
+impl Parse for FunctionComponent {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let parsed: Item = input.parse()?;
+
+        match parsed {
+            Item::Fn(func) => {
+                let ItemFn {
+                    attrs,
+                    vis,
+                    sig,
+                    block,
+                } = func;
+
+                if !sig.generics.params.is_empty() {
+                    // TODO maybe find a way to handle those
```

Comment:
Can we even fully support generics? I'm not sure if `FunctionProvider::run` would accept them as is.

Reply:
Yes, the following input:
```rust
fn foo<P>(_props: &P) -> Html where P: Properties + PartialEq {
    todo!()
}
```

Can be turned into this:
```rust
// generate a tuple struct with a phantom data field for each generic parameter
struct Foo<P>(::std::marker::PhantomData<P>);

// we only need the `where` guard on the trait implementation
impl<P> FunctionProvider for Foo<P> where P: PartialEq + Properties {
    type TProps = P;

    fn run(_props: &P) -> Html {
        todo!()
    }
}

// because this is a type alias we don't need the guards here
type FooComponent<P> = FunctionComponent<Foo<P>>;
```

This can then be used like this you'd expect:
```rust
html! { <FooComponent<()> /> }
```

Reply:
The only problem is that because Rust allows us to use tuple structs with a function-like syntax we can no longer give the struct the same name as the function. I didn't consider this back when I suggested this...

Reply:
I think we should leave this for later. I can't think of a situation where generics wouldn't be needed for function components. What do you think?

Reply:
I think we should do that in a follow up PR though. The code review here is pretty much finished and taking it for another spin would just cause more work.